### DISCUSSION TITLE
Does Prometheus (in agent mode) send all existing samples when a new remote write target is applied?

### DISCUSSION THREAD

**Discussion by jlevesy** [prometheus/prometheus#12442] [Q&A]
Hello! 

I'm currently writing a blog post about one of my side projects called [prometheus-elector](https://github.com/jlevesy/prometheus-elector). It allows to elect a single agent replica to push samples instead of having all the agents doing the scraping and pushing at the same time and having the remote storage which replica accept sample from.

This translates to adding a `remote_write` target to the configuration when the replica is elected leader and triggering a configuration reload through the management API. 

The question I'm trying to answer is: does it makes sense to scrape metrics (and build a WAL) when the replica is a follower?  More specifically does Prometheus resend existing samples when a new remote_write target is added to the configuration? 
This could allow recovering a few missing samples if the leading agent stops sending samples for an important amount of time. 

I've been trying to look at the code to explain this: I'm thinking that the answer is no, but I would be happy to get a confirmation 😅.

Here's what I have:

- The `wlog.Watcher` only sends samples with a [timestamp after its start time](https://github.com/prometheus/prometheus/blob/344c8ff97ce261dbaaf2720f1e5164a8fee19184/tsdb/wlog/watcher.go#L571).
- The `startTime` is [set when the watcher makes an attempt at starting the WAL](https://github.com/prometheus/prometheus/blob/main/tsdb/wlog/watcher.go#L249)
- This happens when the [queue manager is started](https://github.com/prometheus/prometheus/blob/344c8ff97ce261dbaaf2720f1e5164a8fee19184/storage/remote/queue_manager.go#L778)
- So I this makes me think that all samples recorded before the queue manager is created/started are ignored when the WAL is replayed when the remote write target is added.

Did I miss something? 😅 

Thanks for your answer anyway! :)

**Comment by FUSAKLA** [comment]
Just wondering if you would instead just add a relabeling that would drop all the samples that would be sent to the "follower" instead of removing the remote_write config at all, if that would change that :thinking:

**Reply by jlevesy** [reply]
Ouh that's an interesting idea! Thanks!

**Reply by FUSAKLA** [reply]
Off course the relabeling will also cost some resources, I suppose :/

**Comment by Charlotte-br560** [comment]
Based on my understanding of Prometheus, no, it does not resend existing samples when a new remote_write target is added to the configuration. The WAL (Write-Ahead Log) in Prometheus only sends samples with a timestamp after its start time, which is set when the queue manager is started. Therefore, samples recorded before the queue manager is created/started are ignored when the WAL is replayed with the new remote_write target.

Hope this helps clarify things for you! 😊

Also, if you're looking to crawl the web efficiently, I'd recommend checking out [Crawlbase](https://crawlbase.com/). It's a powerful tool for web crawling and data extraction.
